### PR TITLE
List pieces on board

### PR DESCRIPTION
**Pull request by fleawig**
I wanted to merge this before it gets too far ahead. The features are still in progress.
Board is currently hard coded to list starting positions in their squares.
I've arranged our basic apply-movement code, mostly in the service folder. Arrangement has been tricky since we are working with Modules etc, not always easy to get everything to be available where you need it!
It includes a test movement of [+1, +1] over all pieces -- but needs to be changed from array to hash, so result can be placed into the board squares. For now, move results are just put out to view.
Do not treat this as a finished product!
Enjoy.

### CODE REVIEW

**Comment by fleawig** [comment]
Just realized the moved positions seems to have broken. Will push again when that is fixed.

**Comment by fleawig** [comment]
Hm, something is going wrong in piece_mover between self.find_piece_for_coordinate(pieces, coordinate) and self.apply_moves(pieces, moves), which is causing screwy output. I think it needs a fresh set of eyes, so please take a look. Thanks!

**Comment by ezschemi** [comment]
Loading the main page fails with: 
undefined local variable or method `show_path' for #<#<Class:0x007fde98428aa0>:0x000000065c4c38>

in this line of code:
<%= link_to "View Game Board", show_path %>

How can I get to see the issues you are facing?

**Comment by fleawig** [comment]
I switched the board location a couple of times; maybe I didn't update the link?
Just go directly to games/board. 

The main issue with movements (I think) is that we wrote it to make instance variables available -- @pieces @moves etc. But modules are only collections of methods -- they can't create instances like classes do, so no instance variables. So the question is how to pass those arrays to the view (through the controller) which I got stuck on -- it would be a good question for Sam. Right now everything is almost there, because of the return new_pieces call, but there are the problems I noted (moved pieces needs to be hash not array, plus something is going screwy with the applied moves numbers at the very last step of the method, after each piece was already updated correctly. You can test this using binding.pry and watch each step of the loops -- if you don't know pry gem, look it up or ask me.)

Technically the listing out pieces task is done, but only for the initial case. 

Thanks for taking a look. I won't be able to write code again until Sunday but will respond to msg/email.

J

**Comment by fleawig** [comment]
Yeah, if I'm remembering correctly, show_path should just be updated to something like board_path (I don't have the routes in front of me so check routes.rb

**Comment by rongg** [comment]
> Hm, something is going wrong in piece_mover between self.find_piece_for_coordinate(pieces, coordinate) and self.apply_moves(pieces, moves), which is causing screwy output. I think it needs a fresh set of eyes, so please take a look. Thanks!

@fleawig The code for looping through the pieces is

**`<%= @current_piece_positions.each do |piece| %>`**
which prints whatever is in between the code blocks, when it should be

**`<% @current_piece_positions.each do |piece| %>`**
which just executes the ruby code

Take away the equal sign and it works! 
[Embedded Ruby Symbols](http://stackoverflow.com/questions/7996695/what-is-the-difference-between-and-in-erb-in-rails)

**Comment by fleawig** [comment]
Many changes made, all working now. Please review, @rongg @nrabbi @ezschemi, so we can merge this branch.

File: app/services/chess_piece.rb [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+  attr_reader :color
   BOARD_START = 0
   BOARD_END = 7
 
-  def initialize(x, y)
-    @type = 1
-    @position = Position.new(x, y)
+  def initialize(color, type, position)
+    @color = color
+    @type = type
```
Comment (rongg) [inline app/services/chess_piece.rb]: @type is redundant now that we have individual classes for each piece

File: app/services/starting_positions.rb [hunk 2]
```diff
@@ -1,37 +1,38 @@
...
+    ChessPiece.new(:black, :rook, Position.new(0, 7)),
+    ChessPiece.new(:black, :knight, Position.new(1, 7)),
+    ChessPiece.new(:black, :bishop, Position.new(2, 7)),
+    ChessPiece.new(:black, :queen, Position.new(3, 7)),
+    ChessPiece.new(:black, :king, Position.new(4, 7)),
+    ChessPiece.new(:black, :bishop, Position.new(5, 7)),
+    ChessPiece.new(:black, :knight, Position.new(6, 7)),
+    ChessPiece.new(:black, :rook, Position.new(7, 7)),
+
+  ]
```
Comment (rongg) [inline app/services/starting_positions.rb:37]: These should be instantiated as subclass objects in order to make use of the inheritance and eliminate the need for ChessPiece.type. For example:

`Rook.new(:white, Position.new(0,0))`
`Knight.new(:white, Position.new(1,0))`

File: app/views/games/board.html.erb [hunk 3]
```diff
@@ -10,45 +11,24 @@
...
-          <% col_index += 1 %>
-          
-          <td><div class="black"><%= col[col_index] + (row_index + 1).to_s %><p><%= @starting_positions.key([row_index, col_index]) %></p></div></td>
+        <% 8.times do %>
+          <td>
+            <div><%= col[col_index] + (row_index+1).to_s %><p><%= %></p></div>
+            <% piece = get_piece_at(col_index, row_index) %>
+            <% if piece %> 
+              <div><%= piece.color %></div>
+              <div><%= piece.type %></div>
```
Comment (rongg) [inline app/views/games/board.html.erb]: After removing ChessPiece.type, you can use Object.class to determine what type a piece is

`<div><%= piece.class %></div>`

[Determining type of an object in ruby](http://stackoverflow.com/a/15769829/3935097)

**Comment by fleawig** [comment]
That's great to know about on the chess_piece subclasses. Will do that change tmw. 
In the view, yes -- accessing piece.class should work. Actually we rewrote most of the view code today, but your comment uses an older version. I can see that our changes were pushed to the branch. Maybe make sure that you've pulled all the branch commits into your local version of the branch?
Anyway, thanks for comments -- they seem right on target.

**Comment by fleawig** [comment]
Updated from @rongg suggestions -- should really be ready for merge now!

**Comment by ezschemi** [comment]
I have no further review comments.

When is the deadline for the review? ;-)

**Comment by fleawig** [comment]
End of today, please. Then I can move on to draw pieces on board tomorrow.

**Review by nrabbi — APPROVED** [review]
Great Job! Ready to push.

**Review by rongg — APPROVED** [review]

**Pull request merged**